### PR TITLE
feat(source): add decoded preview payloads

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary
- add decoded preview payloads to `ReadSourceResponse` while preserving the canonical raw event envelopes
- teach the source preview service to JSON-decode payload bytes into a structured preview view
- cover the new preview fields in bootstrap and sourceops tests so CLI/HTTP/Connect all expose readable source payloads

## Testing
- make verify
- ./bin/cerebro source read github owner=writer repo=cerebro state=all per_page=1
- GET /sources/github/read?owner=writer&repo=cerebro&state=all&per_page=1

### CODE REVIEW

**Comment by jonathanhaaswriter** [comment]
@droid Please perform a deep code review and security review of this PR. Leave inline comments only for high-confidence, actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review this PR after the latest fixes. Focus on whether the prior code review and security findings are resolved, and call out any remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed conflict-resolution fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
### Droid sweep status

Latest sweep: `2026-04-29 22:09 UTC`.

- Unresolved non-outdated review threads: `0`
- CI `verify`: `SUCCESS`
- `droid-review`: `not present`

I’ll keep watching for new review findings/check failures and address them as they appear.

**Comment by jonathanhaaswriter** [comment]
Re-audited the live unresolved thread (including outdated) against the current branch tip. `SourcePreviewEvent` now carries only `event_id` plus decoded payload metadata, and preview data is returned as a parallel `preview_events` array without embedding a second `EventEnvelope`, so no code change was needed. Validation run: none locally during this audit; existing branch `verify` is green. Remaining unresolved threads after this audit: 0.

**Comment by jonathanhaaswriter** [comment]
Droid sweep audit complete (final pass)

- Re-scanned all live review threads, including outdated ones
- Verified each remaining thread against the current branch tip
- Resolved every actionable thread (or fixed it inline and pushed)
- 0 unresolved review threads across the whole open-PR stack at this time

PR #494 (fix/feedback-main-followup) and PR #495 (feat/droid-creation-workflows) received additional fixes for SSRF (DNS-rebind) and workflow branch-name validation in this pass.

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid please harden this PR end-to-end:

1. Re-review the diff for residual bugs, races, edge cases, security gaps, and untested paths.
2. Apply the smallest correct fixes for any high-confidence issues you find.
3. Strengthen test coverage: add regression tests for every prior review finding on this PR plus the new behavior, including failure paths, boundary inputs, concurrency, and not-found / typed-nil interface guards.
4. Run focused tests and `make verify`. If validators fail, fix until green.
5. Commit and push the fixes to this PR's branch (or, if you lack push permission in this workflow, leave concrete inline suggestions and a summary comment).

Do not change the PR scope, do not modify the default branch, and do not merge.

**Comment by jonathanhaaswriter** [comment]
@droid review

File: sources/github/source.go [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+	return client, settings, nil
+}
+
+// validateBaseURL rejects base URLs whose host targets loopback, private RFC1918, or link-local
+// addresses, or that resolve through DNS to such addresses. This prevents an SSRF where a caller
+// supplies a base_url pointing at internal infrastructure reachable from the cerebro process.
+//
+// When allowLoopback is true (intended for `httptest.Server` based tests only) loopback addresses
+// are permitted. Private/link-local addresses remain rejected regardless.
+func validateBaseURL(ctx context.Context, raw string, allowLoopback bool) error {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/github/source.go]: Fixed in bad98a122 by installing a safe `DialContext` that re-resolves, rejects unsafe IPs at dial time, and dials the validated public IP; added regressions for private-IP rebinding and IP pinning.

**Comment by jonathanhaaswriter** [comment]
@droid review

Please re-review the decoded preview payloads PR deeply, focusing on the new structured view layered on the canonical raw envelope:

1. **Decoder safety**: verify JSON decoding is bounded (max body size, max depth, max field count), uses `DisallowUnknownFields` only where appropriate, and rejects malformed input as `InvalidArgument` rather than returning partial structures.
2. **Memory and CPU bounds**: confirm decode does not allocate unbounded slices/maps for hostile inputs (e.g., deeply nested arrays, gigantic strings, integer overflow in lengths) and that timeouts cover the decode step on slow upstreams.
3. **Raw vs decoded equivalence**: ensure the canonical raw envelope remains the source of truth, that decoded payloads are derived deterministically from it, and that consumers cannot rely on decoded fields that diverge from the raw bytes.
4. **Information disclosure**: check decoded views do not unintentionally surface fields previously redacted in the raw envelope (tokens, headers, secrets, internal IDs) and that error messages from decode failures don’t leak payload contents.
5. **Schema/forward-compat**: verify proto/JSON shape for `preview_events[*].event` and `event_id` stays compatible with PRs 444/446/447 and SDK consumers, and that adding decoded fields is purely additive.
6. **Determinism and ordering**: confirm field ordering, map iteration, and array stability are deterministic across runs and platforms, including how decoded snippets are truncated for display.
7. **Failure-mode coverage**: ensure decode failures, partial decode, and unsupported event kinds are surfaced predictably (preserving the raw envelope) without dropping events or panicking.
